File: srpc/common-rpc.go
package srpc

import (
	"context"
	"io"
	"sync/atomic"

	"github.com/aperturerobotics/starpc/internal/contextutil"
	"github.com/aperturerobotics/util/broadcast"
	"github.com/pkg/errors"
)

// commonRPC contains common logic between server/client rpc.
type commonRPC struct {
	// ctx is the RPC context, canceled when the RPC is canceled.
	ctx context.Context
	// ctxCancel cancels ctx.
	ctxCancel context.CancelFunc
	// ctxCanceled tracks whether ctxCancel has already been called.
	ctxCanceled atomic.Bool
	// service is the rpc service
	service string
	// method is the rpc method
	method string
	// localCompleted tracks if we have sent a completion or cancel locally.
	// note: not guarded by bcast
	localCompleted atomic.Bool
	// bcast guards below fields
	bcast broadcast.Broadcast
	// writer is the writer to write messages to
	writer PacketWriter
	// writerClosed is set after writer has been closed locally.
	writerClosed bool
	// localCompleting is set while the local handler is publishing its terminal
	// packet and closing the writer.
	localCompleting bool
	// localActive is set while the local handler goroutine may still be inside
	// user code. Resource owners use Wait as a lifetime barrier, so cancellation
	// must not make Wait return while a handler can still touch mux-owned state.
	localActive bool
	// localDone is set after the local handler has completed normally.
	localDone bool
	// dataQueue contains incoming data packets.
	// note: packets may be len() == 0
	dataQueue [][]byte
	// dataClosed is a flag set after dataQueue is closed.
	// controlled by HandlePacket.
	dataClosed bool
	// remoteErr is an error set by the remote.
	remoteErr error
	// remoteCanceled distinguishes a received CallCancel from transport loss.
	remoteCanceled bool
	// remoteCompleted is set only by an explicit remote CallData completion.
	remoteCompleted bool
	// remoteTerminal is the first valid remote terminal.
	remoteTerminal TerminalKind
	// remoteTerminalSet records whether remoteTerminal is valid.
	remoteTerminalSet bool
}

// initCommonRPC initializes the commonRPC.
func initCommonRPC(ctx context.Context, rpc *commonRPC) {
	rpc.ctx, rpc.ctxCancel = contextutil.WithCancel(ctx)
}

func (c *commonRPC) cancelContext() {
	if c.ctxCanceled.Swap(true) {
		return
	}
	c.ctxCancel()
}

// Context is canceled when the rpc has finished.
func (c *commonRPC) Context() context.Context {
	return c.ctx
}

// Wait waits for the RPC to finish (remote end closed the stream).
func (c *commonRPC) Wait(ctx context.Context) error {
	for {
		var err error
		var waitCh <-chan struct{}
		var rpcCanceled bool
		var localDone bool
		locked := c.bcast.Lock()
		err = c.remoteErr
		rpcCanceled = c.ctx.Err() != nil
		// A canceled stream tells the handler to stop, but it is not proof that
		// the handler has returned. Keep waiting while localActive is true so a
		// caller that releases resources after Wait cannot race in-flight user
		// code still running on the canceled Stream context.
		if c.localActive {
			waitCh = locked.WaitCh()
			locked.Unlock()

			select {
			case <-ctx.Done():
				return context.Canceled
			case <-waitCh:
				continue
			}
		}
		localDone = c.localDone
		if err == nil && !rpcCanceled && !localDone {
			waitCh = locked.WaitCh()
		}
		locked.Unlock()

		if err != nil {
			return err
		}
		if localDone {
			return nil
		}
		if rpcCanceled {
			return context.Canceled
		}

		select {
		case <-ctx.Done():
			return context.Canceled
		case <-waitCh:
		}
	}
}

// WaitTerminal waits for and classifies the remote terminal of a held unary
// invocation.
func (c *commonRPC) WaitTerminal(ownerCtx context.Context) (TerminalKind, error) {
	var ownerDone bool
	for {
		locked := c.bcast.Lock()
		if c.remoteTerminalSet {
			terminal := c.remoteTerminal
			locked.Unlock()
			return terminal, nil
		}
		if ownerDone {
			err := ownerCtx.Err()
			locked.Unlock()
			return TerminalAbandoned, err
		}
		waitCh := locked.WaitCh()
		locked.Unlock()

		select {
		case <-ownerCtx.Done():
			ownerDone = true
		case <-waitCh:
		}
	}
}

// receiptTerminalKind returns the first valid remote terminal.
func (c *commonRPC) receiptTerminalKind() (TerminalKind, bool) {
	locked := c.bcast.Lock()
	terminal, ok := c.remoteTerminal, c.remoteTerminalSet
	locked.Unlock()
	return terminal, ok
}

// ReadOne reads a single message and returns.
//
// returns io.EOF if the stream ended without a packet.
func (c *commonRPC) ReadOne() ([]byte, error) {
	var ctxDone bool
	for {
		var waitCh <-chan struct{}
		locked := c.bcast.Lock()
		if ctxDone && !c.dataClosed {
			// context must have been canceled locally
			writer := c.closeLocked(&locked)
			locked.Unlock()
			if writer != nil {
				_ = writer.Close()
			}
			return nil, context.Canceled
		}

		if len(c.dataQueue) != 0 {
			msg := c.dataQueue[0]
			c.dataQueue[0] = nil
			c.dataQueue = c.dataQueue[1:]
			locked.Unlock()
			return msg, nil
		}

		if c.dataClosed || c.remoteErr != nil {
			err := c.remoteErr
			if err == nil {
				err = io.EOF
			}
			locked.Unlock()
			return nil, err
		}

		waitCh = locked.WaitCh()
		locked.Unlock()

		select {
		case <-c.ctx.Done():
			ctxDone = true
		case <-waitCh:
		}
	}
}

// WriteCallData writes a call data packet.
func (c *commonRPC) WriteCallData(data []byte, dataIsZero, complete bool, err error) error {
	if complete || err != nil {
		if c.localCompleted.Swap(true) {
			// Repeated completion is an idempotent no-op.
			if complete && len(data) == 0 && !dataIsZero {
				return nil
			}
			return ErrCompleted
		}
	} else if c.localCompleted.Load() {
		return ErrCompleted
	}

	outPkt := NewCallDataPacket(data, len(data) == 0 && dataIsZero, complete, err)
	return c.writer.WritePacket(outPkt)
}

// HandleStreamClose handles the incoming stream closing w/ optional error.
func (c *commonRPC) HandleStreamClose(closeErr error) {
	locked := c.bcast.Lock()
	writer := c.handleStreamCloseLocked(&locked, closeErr)
	locked.Unlock()
	if writer != nil {
		_ = writer.Close()
	}
}

// HandleCallCancel handles the call cancel packet.
func (c *commonRPC) HandleCallCancel() error {
	locked := c.bcast.Lock()
	if (!c.dataClosed || !c.writerClosed) && !c.remoteTerminalSet {
		c.remoteCanceled = true
	}
	writer := c.handleStreamCloseLocked(&locked, context.Canceled)
	locked.Unlock()
	if writer != nil {
		_ = writer.Close()
	}
	return nil
}

// HandleCallData handles the call data packet.
func (c *commonRPC) HandleCallData(pkt *CallData) error {
	var err error
	locked := c.bcast.Lock()
	if c.dataClosed {
		// If the packet is just indicating the call is complete, ignore it.
		if !pkt.GetComplete() {
			// Otherwise, return ErrCompleted (unexpected packet).
			err = ErrCompleted
		}
		locked.Unlock()
		return err
	}

	if data := pkt.GetData(); len(data) != 0 || pkt.GetDataIsZero() {
		c.dataQueue = append(c.dataQueue, data)
	}

	pktErr := pkt.GetError()
	complete := pkt.GetComplete()
	if len(pktErr) != 0 {
		complete = true
		if c.remoteErr == nil {
			c.remoteErr = errors.New(pktErr)
		}
	}

	if complete {
		c.dataClosed = true
		if len(pktErr) == 0 {
			if c.recordRemoteTerminalLocked(TerminalCommitted) {
				c.remoteCompleted = true
			}
		} else {
			c.recordRemoteTerminalLocked(TerminalLost)
		}
	}

	locked.Broadcast()
	locked.Unlock()

	return err
}

func (c *commonRPC) recordRemoteTerminalLocked(kind TerminalKind) bool {
	if c.remoteTerminalSet {
		return false
	}
	c.remoteTerminal = kind
	c.remoteTerminalSet = true
	return true
}

func (c *commonRPC) handleStreamCloseLocked(
	locked *broadcast.Locked,
	closeErr error,
) PacketWriter {
	if c.dataClosed && c.writerClosed {
		return nil
	}
	normalRemoteCloseAfterLocalComplete := closeErr == nil && (c.localCompleting || c.localDone)
	if closeErr != nil && c.remoteErr == nil {
		c.remoteErr = closeErr
	}
	if !normalRemoteCloseAfterLocalComplete && !c.remoteTerminalSet {
		terminal := TerminalClosed
		if closeErr != nil {
			terminal = TerminalLost
			if c.remoteCanceled {
				terminal = TerminalCanceled
			}
		}
		c.recordRemoteTerminalLocked(terminal)
	}
	c.dataClosed = true
	if !normalRemoteCloseAfterLocalComplete {
		c.cancelContext()
		writer := c.closeWriterLocked()
		locked.Broadcast()
		return writer
	}
	locked.Broadcast()
	return nil
}

// WriteCallCancel writes a call cancel packet.
func (c *commonRPC) WriteCallCancel() error {
	// Use atomic swap to check and set completion atomically
	if c.localCompleted.Swap(true) {
		return ErrCompleted
	}

	return c.writer.WritePacket(NewCallCancelPacket())
}

// closeLocked releases resources held by the RPC.
func (c *commonRPC) closeLocked(locked *broadcast.Locked) PacketWriter {
	c.dataClosed = true
	c.localCompleted.Store(true)
	if c.remoteErr == nil {
		c.remoteErr = context.Canceled
	}
	writer := c.closeWriterLocked()
	locked.Broadcast()
	c.cancelContext()
	return writer
}

func (c *commonRPC) closeWriterLocked() PacketWriter {
	if c.writerClosed || c.writer == nil {
		return nil
	}
	c.writerClosed = true
	return c.writer
}

func (c *commonRPC) beginLocalCompletion() {
	locked := c.bcast.Lock()
	c.localCompleted.Store(true)
	c.localCompleting = true
	locked.Unlock()
}

func (c *commonRPC) finishLocalCompletion() {
	locked := c.bcast.Lock()
	c.localCompleted.Store(true)
	writer := c.closeWriterLocked()
	locked.Unlock()
	if writer != nil {
		_ = writer.Close()
	}
	locked = c.bcast.Lock()
	c.localCompleting = false
	c.localActive = false
	c.localDone = true
	locked.Broadcast()
	locked.Unlock()
}

File: srpc/common-rpc.cpp
//go:build deps_only

#include "common-rpc.hpp"

#include "packet.hpp"
#include "rpcproto.pb.h"

namespace starpc {
namespace {

// These bounds contain an unconsumed stream, including zero-length messages.
constexpr size_t kMaximumQueuedBytes = 4 * 1024 * 1024;
constexpr size_t kMaximumQueuedMessages = 1024;

} // namespace

CommonRPC::CommonRPC() = default;
CommonRPC::~CommonRPC() = default;

void CommonRPC::Cancel() {
  {
    std::lock_guard lock(state_mutex_);
    canceled_ = true;
    local_completed_ = true;
    data_queue_.clear();
    queued_bytes_ = 0;
  }
  state_changed_.notify_all();
  stop_source_.request_stop();
  CloseWriter();
}

bool CommonRPC::IsCanceled() const {
  std::lock_guard lock(state_mutex_);
  return canceled_;
}

std::string CommonRPC::RemoteErrorMessage() const {
  std::lock_guard lock(state_mutex_);
  return remote_error_message_;
}

Error CommonRPC::ReadOne(std::string *out) {
  std::unique_lock lock(state_mutex_);
  state_changed_.wait(lock, [this] { return canceled_ || data_closed_ || !data_queue_.empty(); });
  if (canceled_)
    return remote_error_ != Error::OK ? remote_error_ : Error::Canceled;
  if (!data_queue_.empty()) {
    *out = std::move(data_queue_.front());
    queued_bytes_ -= out->size();
    data_queue_.pop_front();
    return Error::OK;
  }
  return remote_error_ != Error::OK ? remote_error_ : Error::EOF_;
}

Error CommonRPC::WriteCallData(const std::string &data, bool data_is_zero, bool complete,
                               Error err) {
  std::unique_lock writing(write_mutex_);
  PacketWriter *writer;
  {
    std::lock_guard lock(state_mutex_);
    if (canceled_)
      return Error::Canceled;
    if (local_completed_) {
      return complete && data.empty() && !data_is_zero ? Error::OK : Error::Completed;
    }
    if (writer_ == nullptr)
      return Error::NilWriter;
    if (writer_closed_)
      return Error::Completed;
    if (complete || err != Error::OK)
      local_completed_ = true;
    writer = writer_;
  }

  const auto packet = NewCallDataPacket(data, data.empty() && data_is_zero, complete, err);
  const auto written = writer->WritePacket(*packet);
  writing.unlock();
  if (written != Error::OK)
    HandleStreamClose(written);
  return written;
}

void CommonRPC::HandleStreamClose(Error close_err) {
  {
    std::lock_guard lock(state_mutex_);
    if (canceled_)
      return;
    if (close_err == Error::EOF_)
      close_err = Error::OK;
    if (remote_error_ == Error::OK && close_err != Error::OK)
      remote_error_ = close_err;
    if (remote_error_ == Error::OK && !remote_completed_ && !local_completing_) {
      remote_error_ = Error::ClosedBeforeCompletion;
    }
    data_closed_ = true;
  }
  state_changed_.notify_all();
  stop_source_.request_stop();
  CloseWriter();
}

Error CommonRPC::HandleCallCancel() {
  Cancel();
  return Error::OK;
}

Error CommonRPC::HandleCallData(const srpc::CallData &pkt) {
  bool overflow = false;
  {
    std::lock_guard lock(state_mutex_);
    if (canceled_)
      return Error::Canceled;
    if (data_closed_)
      return pkt.complete() ? Error::OK : Error::Completed;
    if (!pkt.data().empty() || pkt.data_is_zero()) {
      overflow = data_queue_.size() >= kMaximumQueuedMessages ||
                 pkt.data().size() > kMaximumQueuedBytes - queued_bytes_;
      if (!overflow) {
        data_queue_.push_back(pkt.data());
        queued_bytes_ += pkt.data().size();
      }
    }
    if (!pkt.error().empty()) {
      remote_error_ = Error::RemoteError;
      remote_error_message_ = pkt.error();
    }
    if (pkt.complete() || remote_error_ != Error::OK) {
      data_closed_ = true;
      remote_completed_ = true;
    }
  }
  if (overflow) {
    HandleStreamClose(Error::ResourceExhausted);
    return Error::ResourceExhausted;
  }
  state_changed_.notify_all();
  return Error::OK;
}

Error CommonRPC::WriteCallCancel() {
  std::unique_lock writing(write_mutex_);
  PacketWriter *writer;
  {
    std::lock_guard lock(state_mutex_);
    if (canceled_ || writer_closed_)
      return Error::Canceled;
    if (local_completed_)
      return Error::Completed;
    if (writer_ == nullptr)
      return Error::NilWriter;
    local_completed_ = true;
    writer = writer_;
  }
  const auto written = writer->WritePacket(*NewCallCancelPacket());
  writing.unlock();
  if (written != Error::OK)
    HandleStreamClose(written);
  return written;
}

void CommonRPC::CloseWriter() {
  PacketWriter *writer;
  {
    std::lock_guard lock(state_mutex_);
    if (writer_closed_ || writer_ == nullptr)
      return;
    writer_closed_ = true;
    writer = writer_;
  }
  (void)writer->Close();
}

void CommonRPC::Finish(Error err) {
  {
    std::lock_guard lock(state_mutex_);
    local_completing_ = true;
  }
  (void)WriteCallData("", false, true, err);
  CloseWriter();
  stop_source_.request_stop();
}

} // namespace starpc
